Orange/clustering/clustering.py

```diff
@@ -0,0 +1,86 @@
+import numpy as np
+
+from Orange.data import Table, Instance
+from Orange.preprocess import Continuize, SklImpute
+
+
+def preprocess_data(preprocessors, data):
+    if preprocessors is not None:
+        for pp in preprocessors:
+            data = pp(data)
+    return data
+
+
+class ClusteringModel:
+
+    def __init__(self, projector, pre_domain, preprocessors):
+        self.projector = projector
+        self.pre_domain = pre_domain
+        self.preprocessors = preprocessors
+        self.labels = projector.labels_
+
+    def __call__(self, data):
+        if isinstance(data, Table):
+            if data.domain is not self.pre_domain:
+                data = data.transform(self.pre_domain)
+            data = preprocess_data(self.preprocessors, data)
+            return self.predict(data.X)
+        elif isinstance(data, Instance):
+            if data.domain is not self.pre_domain:
+                data = Instance(self.pre_domain, data)
+            return self.predict(data.x)
+        elif isinstance(data, np.ndarray):
+            return self.predict(data)
+        else:
+            raise Exception("The type of your data is not supported.")
+
+    def predict(self, X):
+        raise NotImplementedError(
+            "This clustering algorithm does not support predicting.")
+
+
+class Clustering:
+    """
+    ${skldoc}
+    Additional Orange parameters
+
+    preprocessors : list, optional (default = [Continuize(), SklImpute()])
+        An ordered list of preprocessors applied to data before
+        training or testing.
+    """
+    __wraps__ = None
+    preprocessors = [Continuize(), SklImpute()]
+    model_class = None
+
+    def __init__(self, preprocessors, parameters):
+        self.preprocessors = tuple(preprocessors or self.preprocessors)
+        self.params = {k: v for k, v in parameters.items()
+                       if k not in ["self", "preprocessors", "__class__"]}
+
+    def __call__(self, data):
+        return self._fit_predict(data)
+
+    def fit_storage(self, data):
+        raise NotImplementedError
+
+    def fit(self, X: np.ndarray, y: np.ndarray = None):
+        return self.__wraps__(**self.params).fit(X, y)
+
+    def _fit_predict(self, data, return_model=False):
```

Comment:
Didn't we discuss that there should be a separate function to get a model, instead of a function like this, that can return different types of results based on a parameter.
Also I think that function should be public (not start with an underscore).